src/java.desktop/share/classes/com/sun/imageio/plugins/png/PNGImageWriter.java

```diff
@@ -919,17 +920,28 @@ private void encodePass(ImageOutputStream os,
 
         int bitDepth = metadata.IHDR_bitDepth;
         for (int row = minY + yOffset; row < minY + height; row += ySkip) {
-            Rectangle rect = new Rectangle(minX, row, width, 1);
-            Raster ras = image.getData(rect);
+            Raster ras;
+            if (image instanceof BufferedImage bi) {
```

Comment:
Do we really need to run these two blocks in the loop for each iteration?

Reply:
Hi Sergey! The current logic does make it very clear where and how `ras` is initialized, which is nice. The top two initialization blocks (for `BufferedImage`s and tiled images) could in theory be hoisted outside the loop, but the third one cannot (and we might have to jump through some hoops to guarantee to the compiler that the variable is initialized if we start initializing it sometimes in one place and sometimes in another).

AFAICT the `BufferedImage.getRaster()` call is not doing any work (just providing a reference to an instance variable), but the `getTile(...)` call might be doing some extra work, depending on the specific instance type (but in any event will not be copying raster data, which was by far the biggest waste here).